### ISSUE TITLE
false positive for `no-string-based-set-interval` rule only when type checking is enabled

### ISSUE THREAD

**Issue by chrismbarr** [microsoft/tslint-microsoft-contrib#374]
`typescript` version 2.2.2
`ts-lint` version 5.2.0
running via `gulp-tslint` 8.0.0

----

When type checlking is enabled the following code fails with `ERROR: 84:32  no-string-based-set-interval  Forbidden setInterval string parameter: process`.

```ts
const process = () => {
    for (let i = 0; i < queue.length; i++) {
        const qi = queue.splice(0, 1)[0];
        if (qi) {
            sendRequest(qi);
        } else {
            break;
        }
    }
};

const intervalLoop = window.setInterval(process, 100);  // <-- failure reported here
```
This **only fails when rule type checking is enabled** though, even though this rule says it does not require type checking!  No failures here are produced when type checking is not enabled.

However, writing it like this does not produce a failure:
```ts
const intervalLoop = window.setInterval(() => {
    for (let i = 0; i < queue.length; i++) {
        const qi = queue.splice(0, 1)[0];
        if (qi) {
            sendRequest(qi);
        } else {
            break;
        }
    }
}, 100);
```

I'm not sure why... I ended up keeping the re-written version since it's simpler, but this feels like a bug.

**Comment by JoshuaKGoldberg** [comment]
The rule doesn't _require_ type checking, but it will take advantage of a TypeScript program if provided one. This must be a bug specifically in that optional logic.

Scanning through the code, `noStringBasedSetIntervalRule.ts` uses a `NoStringParameterToFunctionCallWalker`, which uses its `isExpressionEvaluatingToFunction`. I'd check there for logic that requires `this.typeChecker`.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #876. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!
